tests/source.rs

```diff
@@ -0,0 +1,234 @@
+        .join("target")
+        .join("rls")
+        .join(&Target::Debug.to_string())
+        .join("save-analysis");
+    fs::create_dir_all(&expected_analysis_dir)?;
+    fs::rename(&save_analysis_dir, &expected_analysis_dir)?;
+
+    let host = AnalysisHost::new(Target::Debug);
+    host.reload(&tempdir, &tempdir, true)?;
+
+    Ok(host)
+}
+
+/// Runs all tests in a given source file.
+fn check(source_file: &Path, host: &AnalysisHost) -> Result<()> {
+    let package_name = source_file.file_stem().unwrap();
+    let json = DocData::new(host, package_name.to_str().unwrap())?
+        .to_json()?;
+    let json = serde_json::from_str(&json)?;
+
+    let source = BufReader::new(File::open(source_file)?);
+    let mut found_test = false;
+    for line in source.lines() {
+        if let Some(test_case) = parse_test(&line?) {
+            let (pointer, regex) = test_case?;
+            run_test(&json, &pointer, &regex)?;
+            if !found_test {
+                found_test = true;
+            }
+        }
+    }
+
+    if !found_test {
+        bail!("found no tests in {}", source_file.display());
+    }
+
+    Ok(())
+}
+
+/// Optionally parses a test case from a single line. If the line contains a test case, returns a
+/// Result containing a tuple of the JSON pointer and the regular expression. If there is no test
+/// case contained in the line, returns `None`.
+fn parse_test(line: &str) -> Option<Result<(String, Regex)>> {
+    lazy_static! {
+        static ref COMMAND_RE: Regex =
+            Regex::new("@has (?P<pointer>[a-z/]+) '(?P<match>.+)'").unwrap();
+    }
+
+    if let Some(caps) = COMMAND_RE.captures(line) {
+        let regex = match Regex::new(&caps["match"]) {
+            Ok(regex) => regex,
+            Err(err) => return Some(Err(err.into())),
+        };
+        let pointer = caps["pointer"].to_owned();
+        Some(Ok((pointer, regex)))
+    } else {
+        None
+    }
+}
+
+/// Compares a JSON pointer against a Regex.
+fn run_test(json: &serde_json::Value, pointer: &str, regex: &Regex) -> Result<()> {
+    let value = match json.pointer(pointer) {
+        Some(value) => value,
+        None => bail!(ErrorKind::JsonPointer(pointer.to_owned())),
+    };
+
+    let value = value.as_str().ok_or_else(
+        || "the JSON pointer pointed at a type other than string",
+    )?;
+
+    if regex.is_match(&value) {
+        Ok(())
+    } else {
+        bail!(ErrorKind::ValueMismatch(
+            value.to_owned(),
+            regex.as_str().to_owned(),
+        ));
+    }
+}
+
+/// The test that actually checks all the source tests.
+///
+/// Consider generating these tests in the build script to allow parallelism.
+#[test]
+fn source() {
+    let source_dir = Path::new("tests/source");
+    let tempdir = TempDir::new("rustdoc-test").unwrap();
+
+    for source_file in fs::read_dir(source_dir).unwrap() {
+        let source_file = source_file.unwrap();
+        print!(
+            "checking {} ... ",
```

Comment:
We've added identicaf to the source code so you could use it here to make this nice and pretty!